external/fv3fit/fv3fit/reservoir/domain2.py

```diff
@@ -0,0 +1,396 @@
+from typing import Tuple, Optional
+
+import pace.util
+import numpy as np
+import fsspec
+import yaml
+
+
+def _check_feature_dims_consistent(data_shape, feature_shape):
+    n_feature_dims = len(feature_shape)
+    feature_dims = data_shape[-n_feature_dims:]
+    if feature_dims != tuple(feature_shape):
+        raise ValueError(
+            f"Feature dimensions of data {feature_dims}"
+            f" are not consistent with expected: {feature_shape}"
+        )
+
+
+class RankXYDivider:
+    """
+    Base rank divider class for handling subdomains on a tile
+    (with no overlaps).  Useful for breaking into subdomains
+    for smaller training/prediction tasks.
+
+    Args:
+        subdomain_layout: layout describing subdomain grid within the rank
+            ex. [2,2] means the rank is divided into 4 subdomains
+        rank_extent: Shape of the tile (e.g., [48, 48] for C48 grid)
+        z_feature: Optional trailing feature dimension.  Always assumed
+            to be follow the rank_extent dimensions.
+    """
+
+    def __init__(
+        self,
+        subdomain_layout: Tuple[int, int],
+        rank_extent: Tuple[int, int],
+        z_feature: Optional[int] = None,
+    ):
+        if len(subdomain_layout) != 2:
+            raise ValueError("Rank divider only handles 2D subdomain layouts")
+
+        if len(rank_extent) != 2:
+            raise ValueError(
+                "Rank divider only handles 2D rank extents. A feature dimension "
+                "should be included using 'z_feature' and leading dimensions "
+                "should be excluded."
+            )
+
+        self.rank_extent = rank_extent
+        self._extent_for_check = rank_extent
+        self.subdomain_layout = subdomain_layout
+        self.n_subdomains = subdomain_layout[0] * subdomain_layout[1]
+        self._partitioner = pace.util.TilePartitioner(subdomain_layout)
+
+        self._x_rank_extent = self.rank_extent[0]
+        self._y_rank_extent = self.rank_extent[1]
+        self._z_feature = z_feature
+
+        self._check_extent_divisibility()
+        self._x_subdomain_extent = self._x_rank_extent // self.subdomain_layout[0]
+        self._y_subdomain_extent = self._y_rank_extent // self.subdomain_layout[1]
+
+    @property
+    def subdomain_extent(self):
+        return self._x_subdomain_extent, self._y_subdomain_extent
+
+    @property
+    def _rank_extent_for_partitioner(self):
+        # Fed into partitioner for slicing, no overlap should ever be given
+        return self._maybe_append_feature_value(self.rank_extent, self._z_feature)
+
+    @property
+    def _rank_extent_all_features(self):
+        # used for data consistency check, maybe has overlap in another class
+        return self._rank_extent_for_partitioner
+
+    @property
+    def _rank_dims_all_features(self):
```

Comment:
Make public without _ since it's called in the model code

Reply:
I opted to remove the dependence in the adapter instead.